src/domain/mongoose_domain_core.erl

```diff
@@ -0,0 +1,139 @@
+%% Generally, you should not call anything from this module.
+%% Use mongoose_domain_api module instead.
+-module(mongoose_domain_core).
+-include("mongoose_logger.hrl").
+
+-export([start/1, stop/0]).
+-export([start_link/1]).
+-export([get_host_type/1]).
+-export([is_locked/1]).
+
+%% API for tests
+-export([dump/0, restore/1]).
+
+%% API, used by DB module
+-export([remove_all_unlocked/0,
+         upsert_unlocked/2,
+         remove_unlocked/1]).
+
+-export([get_all_locked/0,
+         get_domains_by_host_type/1]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-define(TABLE, ?MODULE).
+
+start(Pairs) ->
+    ChildSpec =
+        {?MODULE,
+         {?MODULE, start_link, [Pairs]},
+         permanent, infinity, worker, [?MODULE]},
+    just_ok(supervisor:start_child(ejabberd_sup, ChildSpec)).
+
+%% For tests
+stop() ->
+    supervisor:terminate_child(ejabberd_sup, ?MODULE),
+    supervisor:delete_child(ejabberd_sup, ?MODULE),
+    ok.
+
+start_link(Pairs) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [Pairs], []).
+
+get_host_type(Domain) ->
+    case ets:lookup(?TABLE, Domain) of
+        [] ->
+            {error, not_found};
+        [{_Domain, HostType, _Locked}] ->
+            {ok, HostType}
+    end.
+
+is_locked(Domain) ->
+    case ets:lookup(?TABLE, Domain) of
+        [{_Domain, _HostType, _Locked = true}] ->
+            true;
+        _ ->
+            false
+    end.
+
+remove_all_unlocked() ->
+    ets:match_delete(?TABLE, {'_', '_', false}).
+
+get_all_locked() ->
+    heads(ets:match(?TABLE, {'$1', '_', true})).
+
+get_domains_by_host_type(HostType) when is_binary(HostType) ->
+    heads(ets:match(?TABLE, {'$1', HostType, '_'})).
+
+heads(List) ->
+    [H || [H|_] <- List].
+
+upsert_unlocked(Domain, HostType) ->
```

Comment:
There should be no support for upsert, we don't want to change host_type/domain pair ever. We want only to add or remove pairs. Nothing else.

Reply:
unsert is just ets:insert. The preconditions are checked by the caller (i.e. db server). Could be replaced with ets:insert_new though.

Reply:
this must not be possible, and it must be clear from the code of this exact module!